x/dualstaking/keeper/delegate.go

```diff
@@ -0,0 +1,563 @@
+	}
+
+	data, _ := json.Marshal(amount)
+	k.unbondingTS.AddTimerByBlockHeight(ctx, timeout, key, data)
+
+	return nil
+}
+
+// finalizeUnbonding is called when the unbond hold period terminated; it extracts
+// the delegation details from the timer key and data and releases the bonder funds
+// back to the delegator.
+func (k Keeper) finalizeUnbonding(ctx sdk.Context, key []byte, data []byte) {
+	delegator, provider, chainID := decodeForTimer(key)
+
+	attrs := slices.Slice(
+		utils.Attribute{Key: "delegator", Value: delegator},
+		utils.Attribute{Key: "provider", Value: provider},
+		utils.Attribute{Key: "chainID", Value: chainID},
+		utils.Attribute{Key: "data", Value: data},
+	)
+
+	var amount sdk.Coin
+	err := json.Unmarshal(data, &amount)
+	if err != nil {
+		utils.LavaFormatError("critical: finalizeBonding failed to decode", err, attrs...)
+	}
+
+	// sanity: delegator address must be valid
+	delegatorAddr, err := sdk.AccAddressFromBech32(delegator)
+	if err != nil {
+		utils.LavaFormatError("critical: finalizeBonding invalid delegator", err, attrs...)
+		return
+	}
+
+	// sanity: provider address must be valid
+	if _, err := sdk.AccAddressFromBech32(provider); err != nil {
+		utils.LavaFormatError("critical: finalizeBonding invalid provider", err, attrs...)
+		return
+	}
+
+	// sanity: saved amount is valid
+	if err := validateCoins(amount); err != nil {
+		utils.LavaFormatError("critical: finalizeBonding invalid amount", err, attrs...)
+		return
+	} else if amount.IsZero() || amount.IsNegative() {
+		utils.LavaFormatError("critical: finalizeBonding zero amount", err, attrs...)
+		return
+	}
+
+	// sanity: verify that BondedPool has enough funds
+	if k.TotalNotBondedTokens(ctx).LT(amount.Amount) {
+		utils.LavaFormatError("critical: finalizeBonding insufficient bonds", err, attrs...)
+		return
+	}
+
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.NotBondedPoolName, delegatorAddr, sdk.NewCoins(amount))
+	if err != nil {
+		utils.LavaFormatError("critical: finalizeBonding failed to transfer", err, attrs...)
+	}
+
+	details := map[string]string{
+		"delegator": delegator,
+		"provider":  provider,
+		"chainID":   chainID,
+		"amount":    amount.String(),
+	}
+
+	utils.LogLavaEvent(ctx, k.Logger(ctx), types.RefundedEventName, details, "Refunded")
+}
+
+// TODO: this is duplicated in x/pairing/keeper/unstaking.go; merge into one call
+func (k Keeper) getUnbondHoldBlocks(ctx sdk.Context, chainID string) uint64 {
+	_, found, providerType := k.specKeeper.IsSpecFoundAndActive(ctx, chainID)
+	if !found {
+		utils.LavaFormatError("critical: failed to get spec for chainID",
+			fmt.Errorf("unknown chainID"),
+			utils.Attribute{Key: "chainID", Value: chainID},
+		)
+	}
+
+	// note: if spec was not found, the default choice is Spec_dynamic == 0
+
+	block := uint64(ctx.BlockHeight())
+	if providerType == spectypes.Spec_static {
+		return k.epochstorageKeeper.UnstakeHoldBlocksStatic(ctx, block)
+	} else {
```

Comment:
remove the else, then we don't need unreached

Reply:
this is originally _your_ code; I just added the comment to clarify it.